### PR TITLE
AntennaAttributionDocumentReporter: Support multiple project reports

### LINKED ISSUE #2754: Support multiple projects in the AntennaAttributionDocumentReporter
See

https://github.com/oss-review-toolkit/ort/blob/225ab7cbae859ae3e81c39db3b5ad2b62002f576/reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt#L77-L80

We should simple create one PDF per project, which is possible since https://github.com/oss-review-toolkit/ort/pull/2735.

### PR DESCRIPTION
**Pull request by bs-ondem**
Initially, the AntennaAttributionDocumentReporter was able to generate
only one report for a single project and failed if there were mutliple
projects.

After 6fa5c9e reporters are allowed to generate multiple reports.
Therefor the AntennaAttributionDocumentReporter iterates over the
projects and creates for each project a own attribution document with
project name and version as file name prefix.

Resolves #2754.

### CODE REVIEW

**Comment by sschuberth** [comment]
> Therefor

More common spelling: "Therefore"

> a own attribution document

"a separate attribution document"

> with project name and version

"with the project's name and version"

> as file name prefix.

Actually, I'd prefer a suffix, so all `attribution-document-*.pdf` files are grouped when sorted alphabetically in some file view.

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt [hunk 1]
```diff
@@ -132,36 +107,54 @@ class AntennaAttributionDocumentReporter : Reporter {
...
-            DocumentValues(rootProject.id.name, rootProject.id.version, projectCopyright)
-        )
+        val projects = input.ortResult.getProjects(omitExcluded = true)
+        val outputFiles = mutableListOf<File>()
+        try {
+            projects.forEach { project ->
+                val dependencies = project.collectDependencies()
+
+                val packages = input.ortResult.getPackages().mapNotNull { (pkg, _) ->
+                    pkg.takeIf { dependencies.contains(pkg.id) }
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:117]: You can write this as `pkg.takeIf { it.id in dependencies }` (`in` is an infix operator for `contains()`).

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt [hunk 2]
```diff
@@ -132,36 +107,54 @@ class AntennaAttributionDocumentReporter : Reporter {
...
-                if (templateId != DEFAULT_TEMPLATE_ID) removeTemplateFromClassPath()
+
+                val workingDir = createTempDir()
+
+                val documentFile = AttributionDocumentGeneratorImpl(
+                    reportFilename,
+                    workingDir,
+                    templateId,
+                    DocumentValues(project.id.name, project.id.version, projectCopyright)
+                ).generate(artifacts)
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:143]: The case distinction for the directory input is now gone!

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt [hunk 3]
```diff
@@ -132,36 +107,54 @@ class AntennaAttributionDocumentReporter : Reporter {
...
 
-        val generator = AttributionDocumentGeneratorImpl(
-            reportFilename,
-            workingDir,
-            templateId,
-            DocumentValues(rootProject.id.name, rootProject.id.version, projectCopyright)
-        )
+        val projects = input.ortResult.getProjects(omitExcluded = true)
+        val outputFiles = mutableListOf<File>()
+        try {
```
[the remark this answers is not in the document]
Comment (bs-ondem) [reply reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:112]: Does this also apply to if and for?

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt (hunk 2 above)
Comment (bs-ondem) [reply reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt]: Missed that during rebase.

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:116]: Basically yes. However, we sometimes make exceptions e.g. for `val`s that are used in a subsequent `if` (and nowhere else), in which case we sometimes omit the blank line between these. As mentioned, this is a rule of thumb only anyway, but I just wanted to mention it.

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt (hunk 2 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt]: I guessed so :wink:

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt [hunk 4]
```diff
@@ -132,36 +110,73 @@ class AntennaAttributionDocumentReporter : Reporter {
...
+                val documentFile = if (templateFiles.size == templateFileNames.size) {
+                    generator.generate(
+                        artifacts,
+                        templateFiles[TEMPLATE_COVER_PDF],
+                        templateFiles[TEMPLATE_COPYRIGHT_PDF],
+                        templateFiles[TEMPLATE_CONTENT_PDF],
+                        templateFiles[TEMPLATE_BACK_PDF]
+                    )
+                } else {
+                    try {
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:158]: The `try` / `finally` should be dropped here at it's already (correctly) done in the end.

File: reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt [hunk 5]
```diff
@@ -132,36 +110,73 @@ class AntennaAttributionDocumentReporter : Reporter {
...
+                        generator.generate(artifacts)
+                    } finally {
+                        if (templateId != DEFAULT_TEMPLATE_ID) removeTemplateFromClassPath()
+                    }
+                }
+
+                // Antenna keeps around temporary files in its working directory, so we cannot just use our output
+                // directory as its working directory, but have to copy the file we are interested in.
+                val outputFile = outputDir.resolve(
+                    "${REPORT_BASE_FILENAME}-${project.id.name}-${project.id.version}.${REPORT_EXTENSION}"
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/AntennaAttributionDocumentReporter.kt:168]: You can omit the curly braces for the `REPORT_*` constants here as they are not separated by characters that are valid for a variable name.

**Pull request merged**